### PR TITLE
Add Serializer to CascadiaSettings

### PR DESCRIPTION
**Pull request by carlos-zamora**
##  Summary of the Pull Request
This adds `ToJson` functions to `Profile`, `GlobalAppSettings`, and `ColorScheme`. They are used in `CascadiaSettings` to completely serialize an instance of the settings model. Thanks to #7923, all of the settings are `std::optional`, and `JsonUtils` only writes out values that are actually populated.

`CascadiaSettings::WriteSettingsToDisk` serializes the current settings and writes them to the settings.json. A backup file is created with your old contents.

#### Limitations:
- all of the color schemes are serialized regardless of them coming from defaults.json or settings.json
- keybindings/actions are copied/pasted

## References
#1564 - Settings UI
TSM Specs (#6904 and #7876)

## PR Checklist
* [x] Tests added/passed

### CODE REVIEW

**Comment by carlos-zamora** [comment]
### Note to Reviewer
Remaining work before this PR is ready for review:
- [ ] the written tests don't actually pass (yet)
- [ ] `CascadiaSettings` can now create a `Json::Value` representation of its structure. I need to add a function that writes that out to settings.json

Wanted to get some early feedback before these last two steps are done 😊

**Comment by carlos-zamora** [comment]
### Questions for Reviewers
- Hidden profiles are removed from `_profiles`. This means that we won't serialize them out. If that happens to be a dynamic profile, we'll add it back in when the file gets loaded again. What should we do here?
- Do we want to create a backup of `settings.json` when we serialize to `settings.json`? Should that be a part of the same function call to write the json out to a file? Or should they be split up, and TSE needs to be responsible for calling both subsequently?

**Comment by DHowett** [comment]
If you know that a branch targets another branch, you may need to not use automerge ;)

File: src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp
Comment (zadjii-msft) [inline src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp:1079]: Just to make sure, "\_all\_ the color schemes" means _including_ Campbell and the other ones in the `defaults.json`, yea? (again, I'm assuming this is something to be sorted out later in "Write settings to JSON/Part 3")

File: src/cascadia/LocalTests_SettingsModel/SerializationTests.cpp
Comment (zadjii-msft) [inline src/cascadia/LocalTests_SettingsModel/SerializationTests.cpp:66]: Protip: if you wrap the string with `R"(`{YOUR STRING HERE}`)"`, you don't need to keep wrapping the contents in `"`s nor do you need to escape the `"` within the string.

See SettingsTest.cpp for examples

File: src/cascadia/LocalTests_SettingsModel/SerializationTests.cpp
Comment (zadjii-msft) [inline src/cascadia/LocalTests_SettingsModel/SerializationTests.cpp:168]: Can we get tests for roundtripping `foreground` (which is weird), `tabColor` (also weird) and maybe also `source`?

File: src/cascadia/LocalTests_SettingsModel/SerializationTests.cpp
Comment (zadjii-msft) [inline src/cascadia/LocalTests_SettingsModel/SerializationTests.cpp:209]: oh god please do the `R"()"` thing here

File: src/cascadia/TerminalSettingsModel/JsonUtils.h
Comment (zadjii-msft) [inline src/cascadia/TerminalSettingsModel/JsonUtils.h:710]: I'm having a hard time wrapping my head around this code - do we have a good test case that we know tests this string/optional behavior?

File: src/cascadia/LocalTests_SettingsModel/SerializationTests.cpp
Comment (zadjii-msft) [inline src/cascadia/LocalTests_SettingsModel/SerializationTests.cpp:272]: TODO: this

**Review by zadjii-msft — CHANGES_REQUESTED** [review]
There are still some TODOs, though, I suppose my concerns aren't otherwise blocking tbh

File: src/cascadia/TerminalSettingsModel/GlobalAppSettings.cpp
Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsModel/GlobalAppSettings.cpp:356]: I think the correct solution here is to add `debugFeaturesDefault` as the final fallback value like this:
```c++
// GlobalAppSettings.h

#ifdef _DEBUG
static constexpr bool debugFeaturesDefault{ true };
#else
static constexpr bool debugFeaturesDefault{ false };
#endif

GETSET_SETTING(bool, DebugFeaturesEnabled, debugFeaturesDefault);
```

But I think I recall @DHowett not wanting that `static constexpr` in the header file? (pinging him because I don't remember, exactly)

File: src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp
Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp:1079]: Yup. It's gonna be beefy. This'll have to be a part of #8100. I'll link the TODO there for easier tracking.

File: src/cascadia/TerminalSettingsModel/JsonUtils.h
Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsModel/JsonUtils.h:710]: We specifically need this for `StartingDirectory`. If you set it to null in the JSON, we store it as the empty string. For ToJson, we know `_startingDirectory` is an hstring. So that means it is optional, and it is a string. So we pass the hstring to the converter (as opposed to the `optional<hstring>`). The empty string gets treated as an empty string again, and we write it out.

This'll be a good one for me to add as a test along with `foreground`, `tabColor` and maybe `source`. That way, we have all of these weird ones accounted for.

Comment (DHowett) [reply src/cascadia/TerminalSettingsModel/JsonUtils.h:710]: I still don't 100% understand why StartingDirectory couldn't have been a `Nullable<hstring>`

File: src/cascadia/TerminalSettingsModel/TerminalSettingsSerializationHelpers.h
Comment (DHowett) [inline src/cascadia/TerminalSettingsModel/TerminalSettingsSerializationHelpers.h:164]: using exceptions for flow control is not ideal, but it's all we've got

File: src/cascadia/TerminalSettingsModel/Profile.cpp
Comment (DHowett) [reply src/cascadia/TerminalSettingsModel/Profile.cpp:600]: yeah, we would not serialize it properly

File: src/cascadia/TerminalSettingsModel/JsonUtils.h
Comment (DHowett) [inline src/cascadia/TerminalSettingsModel/JsonUtils.h:710]: i also don't 100% understand why this had to be an optional instead of you caching the pointer returned by `demand`...

File: src/cascadia/TerminalSettingsModel/JsonUtils.h
Comment (DHowett) [inline src/cascadia/TerminalSettingsModel/JsonUtils.h:716]: i find this deeply objectionable.

File: src/cascadia/TerminalSettingsModel/JsonUtils.h
Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsModel/JsonUtils.h:710]: The second you use `demand`, you are writing out to the Json::Value. Even if you don't pass anything in, it'll at the very least write out `null` because it assumes the json key exists. In our json model, `null` has meaning (jsoncpp lets you determine if you want null to be written out or ignored). So we only want to do a `demand` call when we know we actually want to write a value out.

Adding a similar comment in the code.

Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsModel/JsonUtils.h:710]: Oh wait, I already have that comment on line 739 haha

Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsModel/JsonUtils.h:710]: If I use `GETSET_NULLABLE_SETTING`, we'll try to expose it as an `IReference<hstring>` and that's not allowed :(. I would have to rewrite the getter/setter to make it truly nullable, but this weird-ish workaround works in the meantime. I'm open to some ideas on how to break it down more eloquently though.

File: src/cascadia/TerminalSettingsModel/CascadiaSettings.idl
Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsModel/CascadiaSettings.idl:14]: Renamed to `WriteSettingsToDisk()`. It creates a settings.json.backup file with your old contents.

We can handle providing a specific filepath later when we make that configurable.

File: src/cascadia/LocalTests_SettingsModel/SerializationTests.cpp
Comment (zadjii-msft) [inline src/cascadia/LocalTests_SettingsModel/SerializationTests.cpp:292]: Wait how the heck does this pass if the keybindings aren't getting re-serialized?

File: src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp
Comment (zadjii-msft) [inline src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp:1053]: Is there a TODO to remove this?

**Review by zadjii-msft — APPROVED** [review]
I'll approve this much - I suppose there's a lot more code coming in #8100 et. al. before this needs to feel totally polished, so let's just unblock this and keep going.

File: src/cascadia/LocalTests_SettingsModel/SerializationTests.cpp
Comment (carlos-zamora) [reply src/cascadia/LocalTests_SettingsModel/SerializationTests.cpp:292]: Ah. We actually copy the Json::Value for actions/keybindings. Then when we serialize our settings, we put them back in.

File: src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp
Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp:1053]: Nope. A part of me wonders if we even want to remove this later? This would be a good conversation to have once Settings UI is on par with JSON though.

File: src/cascadia/LocalTests_SettingsModel/SerializationTests.cpp
Comment (miniksa) [inline src/cascadia/LocalTests_SettingsModel/SerializationTests.cpp:25]: FYI, we can do this now. Did you check whether these tests work in our CI or not? I swear I put this module on in CI.

File: src/cascadia/LocalTests_SettingsModel/SerializationTests.cpp
Comment (miniksa) [inline src/cascadia/LocalTests_SettingsModel/SerializationTests.cpp:55]: Oof. Isn't this expensive to keep building this Builder on every call to the method? Shouldn't it be static or something? I know it's test code and we care less... but this seems a particularly hefty cost.

File: src/cascadia/LocalTests_SettingsModel/SerializationTests.cpp
Comment (miniksa) [inline src/cascadia/LocalTests_SettingsModel/SerializationTests.cpp:241]: You funny with your naming.

File: src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp
Comment (miniksa) [reply src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp:1053]: I really like the idea of backups, though I'd like it even better if it were "previousname-datetime.backup" and just kept rolling. Even more so than taking this feature out. Maybe it's configurable? Keep last 10? But I'm a backup nut.

File: src/cascadia/TerminalSettingsModel/GlobalAppSettings.cpp
Comment (miniksa) [reply src/cascadia/TerminalSettingsModel/GlobalAppSettings.cpp:349]: There's something weird about constexprs in the header that I cannot totally recall at this moment. I'm having a vague memory of when I was trying to fix the constexpr static analysis warnings...

**Review by miniksa — APPROVED** [review]

File: src/cascadia/LocalTests_SettingsModel/SerializationTests.cpp
Comment (carlos-zamora) [reply src/cascadia/LocalTests_SettingsModel/SerializationTests.cpp:25]: It looks like none of the helix tests are running.

File: src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp
Comment (zadjii-msft) [reply src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp:1053]: Wait hold up this won't work at all!
1. This needs to be 
```suggestion
    // write backup settings file
    auto backupSettingsPath{ settingsPath + L".backup" };
    _WriteSettings(_userSettingsString, backupSettingsPath);
```
2. If the user saves the settings twice, won't we just blow away their original settings with the one that we re-serialized after the first launch?

**Review by zadjii-msft — CHANGES_REQUESTED** [review]
oh no

File: src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp
Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp:1053]: > 1. This needs to be 

whoops. Fixed (in upcoming commit).

> 2. If the user saves the settings twice, won't we just blow away their original settings with the one that we re-serialized after the first launch?

Yeah, we only hold one backup at a time. We could add multiple backups like Michael was saying ("settings.json-<date><time>.backup"). But won't that take up too much space? (also, how do you get the current date/time?)

File: src/cascadia/TerminalSettingsModel/GlobalAppSettings.cpp
Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsModel/GlobalAppSettings.cpp:349]: Moved it back to the cpp and added a helper function. Sounds like we _know_ it's bad, but can't remember why hahaha.

File: src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp
Comment (zadjii-msft) [reply src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp]: We could always just not write that file _if it exists_, yea? Presumably the first time, we'll back it up, and then we'll only ever touch the settings-ui modified version

Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp]: @zadjii-msft Done. Not sure if there's a faster way to check if a file exists than `CreateFileW()`. A bit annoyed that we'd call it twice (once to check if the file exists, and again to write it), but I'm not too familiar with this stuff.

**Review by zadjii-msft — APPROVED** [review]

**Review by DHowett — COMMENTED** [review]
`CREATE_NEW`: Creates a new file, only if it does not already exist.
If the specified file exists, the function fails and the last-error code is set to `ERROR_FILE_EXISTS` (80).

This is how you would avoid opening the file twice and introducing a race condition.

**Review by DHowett — CHANGES_REQUESTED** [review]
NAK

File: src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp
Comment (DHowett) [inline src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp:1060]: @miniksa why exception handling? this is going to happen on every single save operation after the first, so it doesn't seem particularly exceptional

Comment (miniksa) [reply src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp:1060]: RETURN_LAST_ERROR_IF and HRESULT it so it's not all exceptions? I dunno. I find it weird to make a backup ONCE and then never overwrite it again in the first place.

I feel like the backups should be versioned with datetime stamp or that the backup should be the most recent one, not the oldest one.

Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp:1060]: My concern with creating a string of backups is that we'll be taking up more storage space on every save.

File: src/cascadia/LocalTests_SettingsModel/JsonTestClass.h
Comment (DHowett) [inline src/cascadia/LocalTests_SettingsModel/JsonTestClass.h:29]:
```suggestion
    }
```

File: src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp
Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp:1060]: Discussed offline. These backup files are _tiny_ (7kb). So we'll do a datetime stamp. Once our settings UI and model is good to go, we'll remove this (or make it configurable).

File: src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp
Comment (DHowett) [inline src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp:1054]: definitely don't do this! it'll turn jan 1 at 1:01:01 AM into "202111111" instead of "20210101010101"!

Comment (DHowett) [reply src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp:1054]: you may wanna consider...


```
#include <fmt/chrono.h>
...
std::time_t timeStamp;
std::time(&timeStamp);
xxx = fmt::format(L"{}.{:%Y-%m-%dT%H:%M:%S}.backup", settingsPath, fmt::localtime(timeStamp));
```

(fmt has a strftime time formatter!)

Comment (DHowett) [reply src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp:1054]: though should we consider using `std::chrono::system_clock` to get the time_t?

File: src/cascadia/TerminalSettingsModel/Profile.cpp
Comment (DHowett) [inline src/cascadia/TerminalSettingsModel/Profile.cpp:573]: this comment is no longer necessary, is it?

Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsModel/Profile.cpp]: removed

File: src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp
Comment (DHowett) [inline src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp:1088]: i hope to heck that this doesn't use the json format with the spaces _before_ the colons for some reason

File: src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp
Comment (DHowett) [inline src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp:1106]: umm..

if the user doesn't have a defaults block this is 100% going to crash my dude

Comment (DHowett) [reply src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp:1106]: i'm going to need you to test both this and the legacy profiles layout (array only)

File: src/cascadia/TerminalSettingsModel/GlobalAppSettings.cpp
Comment (DHowett) [reply src/cascadia/TerminalSettingsModel/GlobalAppSettings.cpp:365]: I'm unresolving this conversation because it never explained why we were skipping serializing it and what the user experience is for when they try to round-trip it.

If we are not serializing a setting, we had better put a damn good comment in the code explaining not just _that_ we're skipping it but _WHY_ we're skipping it.

File: src/cascadia/TerminalSettingsModel/Profile.cpp
Comment (DHowett) [inline src/cascadia/TerminalSettingsModel/Profile.cpp:294]: 'scuse me?

**Review by DHowett — CHANGES_REQUESTED** [review]

File: src/cascadia/TerminalSettingsModel/GlobalAppSettings.cpp
Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsModel/GlobalAppSettings.cpp:365]: It's been a while since I looked at this. I'm adding it back in and removing the comment. Since we're having `debugFeaturesDefault` be the system set fallback value, no reason why we should be skipping it.

File: src/cascadia/TerminalSettingsModel/JsonUtils.h
Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsModel/JsonUtils.h]: (Leaving this comment for tracking) #8175 allowed for the removal of this hack.

File: src/cascadia/TerminalSettingsModel/Profile.cpp
Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsModel/Profile.cpp:294]: Removed

File: src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp
Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp:1088]: Yeah. Just switched to a `StreamWriterBuilder`. Looks better now. The _only_ thing that really bugs me, is that everything is written in alphabetical order. So we'll get some global settings written all the way at the bottom.

File: src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp
Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp:1106]: Latest commit verifies that this doesn't crash :) Good catch.

File: src/cascadia/TerminalSettingsModel/Profile.cpp
Comment (DHowett) [reply src/cascadia/TerminalSettingsModel/Profile.cpp]: Why did you remove it? Why had you added it? It seems very important to serialize it, but i’m surprised it wasn’t there before

Don’t just go deleting lines of code because someone commented on them!

Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsModel/Profile.cpp]: Oof. I forgot to run the tests anyways. So, this is interesting. I looked through the history of Profile.cpp and it looks like it was always missing (checked the inheritance and winrt-ification PRs). But, we need this so that the following roundtrip test passes:
```
            {
                "name": "Weird Profile",
                "tabColor": null,
                "foreground": null,
                "source": "local"
            }
```

I think we just always relied on `source` being set by the dynamic profile generator, so we never had to care. But now with inheritance and serialization, we need this? Still a bit weird because really the user shouldn't be setting a `source`, Terminal should. Thoughts? Is this even a valid test?

Comment (DHowett) [reply src/cascadia/TerminalSettingsModel/Profile.cpp]: it was set by "generateStub", if I recall correctly.

If there is a source, it must be present in the user's settings.

Comment (DHowett) [reply src/cascadia/TerminalSettingsModel/Profile.cpp]: because source is part of the primary key we use to match up dynamic profiles.

Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsModel/Profile.cpp:294]: Ok. Then I added it back in.

**Review by DHowett — APPROVED** [review]
Lets do this

**Pull request merged**
